packages/router/src/components/RouteMatch.tsx

```diff
@@ -58,13 +53,7 @@ const TraverseRootComponents = React.memo(
     children,
   }: TraverseRootComponentsProps): React.JSX.Element => {
     if (routes.length > index) {
-      const Parent = React.useMemo(
-        () =>
-          routes[index]?.component as unknown as (
-            props: ParentProps,
-          ) => React.JSX.Element,
-        [],
-      )
+      const Parent = (routes[index] as Route).component
```

Comment:
- The optional chain should not be needed: if route is `undefined` the `<Parent>` below should have throw an error since `undefined` is not a `React.ElementType`
- The `useMemo` has no dependencies, so if the `routes` or `index` prop changes the memo will still have the inital reference which might cause some bug or gc issues

Due to the previous points removing `useMemo` seems the most viable way to resolve the rule of hooks error